pkg/controllers/raycluster_webhook.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"github.com/project-codeflare/codeflare-operator/pkg/config"
+)
+
+// log is for logging in this package.
+var rayclusterlog = logf.Log.WithName("raycluster-resource")
+
+func (r *RayClusterDefaulter) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&rayv1.RayCluster{}).
+		WithDefaulter(&RayClusterDefaulter{
+			Config:                   r.Config,
+			rayDashboardOauthEnabled: r.isRayDashboardOAuthEnabledWebhook(),
+		}).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-ray-io-v1-raycluster,mutating=true,failurePolicy=fail,sideEffects=None,groups=ray.io,resources=rayclusters,verbs=create;update,versions=v1,name=mraycluster.kb.io,admissionReviewVersions=v1
+
+type RayClusterDefaulter struct {
+	Config                   *config.KubeRayConfiguration
+	rayDashboardOauthEnabled bool
+}
+
+var _ webhook.CustomDefaulter = &RayClusterDefaulter{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *RayClusterDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	raycluster := obj.(*rayv1.RayCluster)
+
+	if r.rayDashboardOauthEnabled {
+		rayclusterlog.Info("default", "name", raycluster.Name)
+		// Check and add OAuth proxy if it does not exist.
+		alreadyExists := false
+		for _, container := range raycluster.Spec.HeadGroupSpec.Template.Spec.Containers {
+			if container.Name == "oauth-proxy" {
+				rayclusterlog.Info("OAuth sidecar already exists, no patch needed")
+				alreadyExists = true
+				break // exits the for loop
+			}
+		}
+
+		if !alreadyExists {
+			rayclusterlog.Info("Adding OAuth sidecar container")
+			// definition of the new container
+			newOAuthSidecar := corev1.Container{
```

Comment:
Would it have sense to limit Resources for oauth container?

Reply:
Thanks @sutaakar 
I feel this is a good idea, I'd be unsure of the limits to set, so I'll pull in @astefanutti  :)